File: src/main/java/location/filters/LocationFilter.java
package location.filters;


import location.LocationRecord;

/**
 * Filters records that are inside or outside a given radius around a given point
 */
public class LocationFilter extends RecordFilter {

    // Earth's mean radius in kms
    private static final double earthRad = 6371.0088;

    private double latitude;
    private double longitude;
    private double radius;
    private boolean acceptInside;

    public LocationFilter(double latitude, double longitude, double radius, boolean acceptInside) {
        this.latitude = latitude;
        this.longitude = longitude;
        this.radius = radius;
        this.acceptInside = acceptInside;
    }

    @Override
    protected boolean test(LocationRecord record) {
        double dist = dist(record.getLatitude(), record.getLongitude());
        return acceptInside ? (dist <= radius) : (dist >= radius);
    }

    private double dist(double lat, double lon) {
        double rLat1 = degToRad(lat);
        double rLon1 = degToRad(lon);
        double rLat2 = degToRad(latitude);
        double rLon2 = degToRad(longitude);

        // Half of delta latitude
        double hdLat = (rLat1 - rLat2) / 2;
        // Half of delta longitude
        double hdLon = (rLon1 - rLon2) / 2;

        double a = Math.sin(hdLat) * Math.sin(hdLat) +
                Math.cos(rLat1) * Math.cos(rLat2) * Math.sin(hdLon) * Math.sin(hdLon);
        double c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1.0 - a));

        return c * earthRad;
    }

    private double degToRad(double deg) {
        return deg / 180.0 * Math.PI;
    }

}

File: src/main/kotlin/location/filters/LocationFilter.kt
package location.filters


import location.LocationRecord

/**
 * Filters records that are inside or outside a given radius around a given point
 */
class LocationFilter(private val latitude: Double,
                     private val longitude: Double,
                     private val radius: Double,
                     private val acceptInside: Boolean) : RecordFilter() {

    companion object {
        private const val earthRad = 6371.0088 // Earth's mean radius in kms
    }

    override fun test(record: LocationRecord): Boolean {
        val dist = dist(record.latitude, record.longitude)

        return when {
            acceptInside ->
                dist <= radius
            else ->
                dist >= radius
        }
    }

    private fun dist(lat: Double, lon: Double): Double {
        val rLat1 = degToRad(lat)
        val rLon1 = degToRad(lon)
        val rLat2 = degToRad(latitude)
        val rLon2 = degToRad(longitude)

        // Half of delta latitude
        val hdLat = (rLat1 - rLat2) / 2
        // Half of delta longitude
        val hdLon = (rLon1 - rLon2) / 2

        val a = Math.sin(hdLat) * Math.sin(hdLat) + Math.cos(rLat1) * Math.cos(rLat2) * Math.sin(hdLon) * Math.sin(hdLon)
        val c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1.0 - a))

        return c * earthRad
    }

    private fun degToRad(deg: Double) = deg / 180.0 * Math.PI

}
